substrate/pkg/controller/substrate/vpc.go

```diff
@@ -1,78 +1,74 @@
 	if err != nil {
-		return err
+		return reconcile.Result{}, fmt.Errorf("describing vpc, %w", err)
 	}
-	// vpc doesn't exist create VPC
-	vpcID := ""
-	if vpc == nil || *vpc.VpcId == "" {
-		result, err := v.ec2Client.CreateVpc(&ec2.CreateVpcInput{
-			CidrBlock:         aws.String(substrate.Spec.VPC.CIDR),
-			TagSpecifications: generateEC2Tags(v.resourceName(), substrate.Name),
-		})
-		if err != nil {
-			return fmt.Errorf("creating VPC, %w", err)
-		}
-		logging.FromContext(ctx).Infof("Created VPC %v ID %v", substrate.Name, *result.Vpc.VpcId)
-		vpcID = *result.Vpc.VpcId
-	} else {
-		vpcID = *vpc.VpcId
+	if len(describeVpcsOutput.Vpcs) > 0 {
+		substrate.Status.VPCID = describeVpcsOutput.Vpcs[0].VpcId
+		logging.FromContext(ctx).Infof("Found vpc %s", aws.StringValue(substrate.Status.VPCID))
+		return reconcile.Result{}, nil
 	}
-	substrate.Status.VPCID = aws.String(vpcID)
-	return err
-}
-
-func (v *vpc) Delete(ctx context.Context, substrate *v1alpha1.Substrate) error {
-	vpc, err := v.getVPC(ctx, substrate.Name)
+	createVpcOutput, err := v.ec2Client.CreateVpc(&ec2.CreateVpcInput{
+		CidrBlock:         aws.String(substrate.Spec.VPC.CIDR),
+		TagSpecifications: tagsFor(ec2.ResourceTypeVpc, substrate.Name, substrate.Name),
+	})
 	if err != nil {
-		return err
-	}
-	// vpc doesn't exist, return
-	if vpc == nil || *vpc.VpcId == "" {
-		return nil
+		return reconcile.Result{}, fmt.Errorf("creating VPC, %w", err)
 	}
-	if _, err := v.ec2Client.DeleteVpcWithContext(ctx, &ec2.DeleteVpcInput{
-		VpcId: vpc.VpcId,
-	}); err != nil {
-		return fmt.Errorf("deleting vpc, %w", err)
+	if createVpcOutput.Vpc.VpcId == nil {
+		return reconcile.Result{Requeue: true}, nil
 	}
-	return nil
+	substrate.Status.VPCID = createVpcOutput.Vpc.VpcId
+	logging.FromContext(ctx).Infof("Created vpc %s", aws.StringValue(substrate.Status.VPCID))
+	return reconcile.Result{}, err
 }
 
-func (v *vpc) getVPC(ctx context.Context, identifier string) (*ec2.Vpc, error) {
-	input := &ec2.DescribeVpcsInput{
-		Filters: ec2FilterFor(identifier),
-	}
-	output, err := v.ec2Client.DescribeVpcsWithContext(ctx, input)
+func (v *vpc) Delete(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	describeVpcsOutput, err := v.ec2Client.DescribeVpcsWithContext(ctx, &ec2.DescribeVpcsInput{Filters: filtersFor(substrate.Name)})
 	if err != nil {
-		return nil, fmt.Errorf("describing VPC for %v, err: %w", identifier, err)
+		return reconcile.Result{}, fmt.Errorf("describing vpc, %w", err)
 	}
-	// Check if VPC doesn't exist return no error
-	if len(output.Vpcs) == 0 {
-		return nil, nil
+	if len(describeVpcsOutput.Vpcs) == 0 {
+		return reconcile.Result{}, nil
 	}
-	if len(output.Vpcs) > 1 {
-		return nil, fmt.Errorf("expected to find one VPC, but found %v", len(output.Vpcs))
+	if _, err := v.ec2Client.DeleteVpcWithContext(ctx, &ec2.DeleteVpcInput{VpcId: describeVpcsOutput.Vpcs[0].VpcId}); err != nil {
```

Comment:
we should loop through all vpcs, just in case there are multiple of them created as there can be multiple with same name

Reply:
+1